.github/workflows/lint.yaml

```diff
@@ -1,20 +1,150 @@
+# TODO(zmb3): change the name of this check
+# (it's no longer just Go, but we have to change the required checks before we can rename)
 name: Lint (Go)
 run-name: make lint
 on:
   pull_request:
-    paths-ignore:
-      - "docs/**"
-      - "rfd/**"
-      - "**/*.md*"
+
   merge_group:
-    paths-ignore:
-      - "docs/**"
-      - "rfd/**"
-      - "**/*.md*"
 
 jobs:
-  lint:
-    name: Lint (Go)
+  changes:
+    name: Check for relevant changes
+    runs-on: ubuntu-latest
+    permissions:
+      pull-requests: read
+    outputs:
+      go: ${{ steps.changes.outputs.go }}
+      godeps: ${{ steps.changes.outputs.godeps }}
+      helm: ${{ steps.changes.outputs.helm }}
+      proto: ${{ steps.changes.outputs.proto }}
+      rust: ${{ steps.changes.outputs.rust }}
+      sh: ${{ steps.changes.outputs.sh }}
+      terraform: ${{ steps.changes.outputs.terraform }}
+      web: ${{ steps.changes.outputs.web }}
+    steps:
+      - name: Checkout
+        if: ${{ github.event_name == 'merge_group' }}
+        uses: actions/checkout@v4
+      - uses: dorny/paths-filter@4512585405083f25c027a35db413c2b3b9006d50 # v2.11.1
+        id: changes
+        with:
+          base: ${{ github.event.pull_request.base.ref || github.event.merge_group.base_ref }}
+          ref: ${{ github.event.pull_request.head.ref || github.event.merge_group.head_ref }}
+          filters: |
+            go:
+              - **/*.go
+              - **/go.mod
+              - **/go.sum
+            godeps:
+              - **/go.mod
+              - **/go.sum
+            helm:
+              - examples/chart/**
+            rust:
+              - **/Cargo.lock
+              - **/Cargo.toml
+              - **/*.rs
+            proto:
+              - **/*.proto
+            web:
+              - web/**
+            terraform:
+              - '**.tf'
+              - '**.tf.json'
+              - '**.hcl'
+            sh:
+              - **/*.sh
+              - **/*.sh.tmpl
+
+  lint-terraform:
+    needs: changes
+    if: ${{ needs.changes.outputs.terraform == 'true' }}
+    uses: gravitational/shared-workflows/.github/workflows/terraform-lint.yaml@main
+    permissions:
+      actions: read
+      contents: read
+      pull-requests: write
+      security-events: write
+
+  lint-rust:
+    name: Lint (Rust)
+    needs: changes
+    if: ${{ needs.changes.outputs.rust == 'true' }}
+    runs-on: ubuntu-22.04-16core
+
+    permissions:
+      contents: read
+
+    container:
+      image: ghcr.io/gravitational/teleport-buildbox:teleport14
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        # TODO(codingllama): Consider https://github.com/actions-rs/clippy-check
+      - name: Run Rust linter
+        run: make lint-rust lint-license
+
+  lint-helm:
```

Comment:
This new job is necessary because it used to be run indirectly along with the rust lints in `make lint-no-actions`.